src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,180 @@
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                const loop = new DetachableLoop(this.dbosExec, ro.schedulerConfig.crontab ?? '* * * * *', ro);
+                this.schedLoops.push(loop);
+                this.schedTasks.push(loop.startLoop());
+            }
+        }
+    }
+  
+    async destroyScheduler() {
+        for (const l of this.schedLoops) {
+            l.setStopLoopFlag();
+        }
+        this.schedLoops = [];
+        try {
+           await Promise.allSettled(this.schedTasks);
+        }
+        catch (e) {
+           //  What gets caught here is the loop stopping, which is what we wanted.
+        }
+        this.schedTasks = [];
+    }
+  
+    logRegisteredSchedulerEndpoints() {
+        const logger = this.dbosExec.logger;
+        logger.info("Scheduled endpoints:");
+        this.dbosExec.registeredOperations.forEach((registeredOperation) => {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly scheduledMethod: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.scheduledMethod.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
```

Comment:
Maybe a naive question but can we simply leverage `node-cron` itself to schedule? Then we put our logic to start a workflow as the scheduled cron job.